toolchain/parser/parser_handle_index_expression.cpp

```diff
@@ -0,0 +1,30 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/parser/parser_context.h"
+
+namespace Carbon {
+
+auto ParserHandleIndexExpression(ParserContext& context) -> void {
+  auto state = context.PopState();
+  state.state = ParserState::IndexExpressionFinish;
+  context.PushState(state);
+  context.AddNode(ParseNodeKind::IndexExpressionStart, context.Consume(),
```

Comment:
```suggestion
  context.AddNode(ParseNodeKind::IndexExpressionStart, context.ConsumeChecked(TokenKind::OpenSquareBracket),
```

Please specify the token being consumed so that it can be validated. This is intended to help catch programming mistakes. Not all code is doing this yet, but it's the direction we want to head.